frontend/src/components/widgets/WidgetSpan.svelte

```diff
@@ -76,31 +84,31 @@
 	{#each widgets as component, index}
 		{@const checkboxInput = narrowWidgetProps(component.props, "CheckboxInput")}
 		{#if checkboxInput}
-			<CheckboxInput {...exclude(checkboxInput)} on:checked={({ detail }) => updateLayout(index, detail)} />
+			<CheckboxInput {...exclude(checkboxInput)} on:checked={({ detail }) => commitAndUpdateLayout(index, detail)} />
 		{/if}
 		{@const colorInput = narrowWidgetProps(component.props, "ColorButton")}
 		{#if colorInput}
-			<ColorButton {...exclude(colorInput)} on:value={({ detail }) => updateLayout(index, detail)} />
+			<ColorButton {...exclude(colorInput)} on:value={({ detail }) => updateLayout(index, detail)} on:start={() => commitLayout(index)} />
```

Comment:
Can you please help me understand why we are committing the layout upon the start of the user's dragging? Why doesn't this happen upon completion of the dragging?

Reply:
it also seems unreasonable to me first. After tried myself, I try to understand like this: for undo history, we always need to commit a undo history and keep a snapshot before we begin new changes. Then upon the start of dragging means commit a undo history before user begin new dragging changes. 

Actually we commit the layout of a last status, not a layout changed by this widget. Maybe we could consider change a name of this.

Reply:
But doesn't that mean that we don't save a history step upon completing the drag? So if you change a number from 1 to 10 by dragging its NumberInput slider, that value of 10 never gets saved as a history step? So the user could move on to other things, and if they undo several times, the history jumps back two operations at once (the operation that moved from 1 to 10, and whatever the other operation was adjacent to that).

Reply:
> But doesn't that mean that we don't save a history step upon completing the drag? 

saving a history step upon completing the drag is a responsibility belongs to the next widget's commit_layout or next action that trigger `Document.startTransaction`. That's why common widgets use `updateAndCommitLayout`.

Reply:
> So if you change a number from 1 to 10 by dragging its NumberInput slider, that value of 10 never gets saved as a history step?

if user's next action is to add a path using pen tool. During we adding a path, we will trigger `Document.StartTransanction` then save a history step. (Our existing history related code have handled this properly I think)

Reply:
I see! I think part of my confusion is because I don't have a full understanding of how `StartTransaction`/`AbortTransaction`/`CommitTransaction` work (since the original author of that never documented them). I did test this all out and it seems to work well so I trust that you're right about this. I'll keep going with my code review.

Reply:
`CommitTransaction` is a noop. `StartTransaction` clones the document and stores it in a list. `AbortTransaction` pops a document from the list and makes it current.

Reply:
That's interesting that `CommitTransaction` is a noop.  Is there a reason we have it? Do you think we should we just remove it, then find more intuitive names for the remaining two?

Reply:
Your welcome to modify it.